certgen/certgen_test.go

```diff
@@ -0,0 +1,242 @@
+	if err != nil {
+		t.Fatalf("root and key setup error: %s", err)
+	}
+
+	timeNowUTC = func() time.Time {
+		return time.Date(2019, time.March, 25, 12, 0, 0, 0, time.UTC)
+	}
+
+	for _, test := range tests {
+		t.Run(test.desc, func(t *testing.T) {
+			ca := &CA{
+				SigningCert: root,
+				SigningKey:  rootKey,
+				CertConfig: CertificateConfig{
+					SubjectCommonName:         subjectCommonName,
+					SubjectOrganization:       subjectOrganization,
+					SubjectOrganizationalUnit: subjectOrganizationalUnit,
+					SubjectLocality:           subjectLocality,
+					SubjectCountry:            subjectCountry,
+					SignatureAlgorithm:        signatureAlgorithm,
+					DNSPrefix:                 prefix,
+					NotAfterInterval:          test.notAfterInterval,
+				},
+			}
+
+			cert, err := ca.IssueCertificate()
+			if err != nil {
+				t.Fatalf("error creating certificate: %s", err)
+			}
+
+			// Check the fields that are set in the leaf template are present
+			// and correct.
+
+			if cert.SerialNumber == nil {
+				t.Error("certificate Serial Number is nil")
+			}
+
+			cc := ca.CertConfig
+			// Check the Subject fields.
+			if want := []string{cc.SubjectCountry}; !cmp.Equal(cert.Subject.Country, want) {
+				t.Errorf("certificate Subject Country = %v, want %v", cert.Subject.Country, want)
+			}
+			if want := []string{cc.SubjectOrganization}; !cmp.Equal(cert.Subject.Organization, want) {
+				t.Errorf("certificate Subject Organization = %v, want %v", cert.Subject.Organization, want)
+			}
+			if want := []string{cc.SubjectOrganizationalUnit}; !cmp.Equal(cert.Subject.OrganizationalUnit, want) {
+				t.Errorf("certificate Subject OrganizationalUnit = %v, want %v", cert.Subject.OrganizationalUnit, want)
+			}
+			if want := []string{cc.SubjectLocality}; !cmp.Equal(cert.Subject.Locality, want) {
+				t.Errorf("certificate Subject Locality = %v, want %v", cert.Subject.Locality, want)
+			}
+			if cert.Subject.CommonName != cc.SubjectCommonName {
+				t.Errorf("certificate Subject Common Name = %s, want %s", cert.Subject.CommonName, cc.SubjectCommonName)
+			}
+
+			// Check the validity period fields.
+			if cert.NotBefore.IsZero() {
+				t.Error("certificate NotBefore is the zero time")
+			}
+			if want := cert.NotAfter.Add(-certValidity); want != cert.NotBefore {
+				t.Errorf("certificate NotBefore = %s, want %s (%s before Not After)", cert.NotBefore, want, certValidity)
+			}
+			if cert.NotAfter.IsZero() {
+				t.Error("certificate NotAfter is the zero time")
+			}
+			if cc.NotAfterInterval != nil {
+				// Check that cert.NotAfter is in the NotAfterInterval [Start, End).
+				if cert.NotAfter.Before(cc.NotAfterInterval.Start) || !cc.NotAfterInterval.End.After(cert.NotAfter) {
+					t.Errorf("certificate NotAfter = %s, should be between [%s, %s)", cert.NotAfter, cc.NotAfterInterval.Start, cc.NotAfterInterval.End)
+				}
+			}
+
+			// Check the extension fields.
+			if cert.KeyUsage != x509.KeyUsageDigitalSignature {
```

Comment:
you could assign `got, want` here to avoid repeating the long const names as parameters in the Errorf call.

Reply:
Yup, have changed all of the simpler ones.